File: frontend/src/api/clasPipeline.ts
/**
 * API client for the CLAS pipeline (mrtk relay + mrtk cssr2rtcm3).
 */

const API_BASE = '/api/clas-pipeline';

export interface ReceiverPreset {
  id: string;
  vendor: string;
  model: string;
  label: string;
  relay_input_format: string;
  default_input_baud: number;
  default_output_baud: number;
  notes: string;
}

export interface SerialPort {
  path: string;
  label: string;
}

export interface PipelineStatus {
  state: 'idle' | 'starting' | 'running' | 'stopping' | 'stopped' | 'error';
  relay_state: string;
  cssr_state: string;
  started_at: string | null;
  error_message: string | null;
  bridge_port: number | null;
}

export interface PipelineStartRequest {
  receiver_id: string;
  input_device: string;
  input_baud: number;
  output_device: string;
  output_baud: number;
  bridge_port?: number;
  sbf_record_path?: string | null;
}

async function handle<T>(res: Response): Promise<T> {
  if (!res.ok) {
    const err = await res.json().catch(() => ({ detail: res.statusText }));
    throw new Error(err.detail || 'API request failed');
  }
  return res.json();
}

export async function listReceivers(): Promise<ReceiverPreset[]> {
  return handle(await fetch(`${API_BASE}/receivers`));
}

export async function listSerialPorts(): Promise<SerialPort[]> {
  return handle(await fetch(`${API_BASE}/serial-ports`));
}

export async function startPipeline(req: PipelineStartRequest): Promise<PipelineStatus> {
  return handle(
    await fetch(`${API_BASE}/start`, {
      method: 'POST',
      headers: { 'Content-Type': 'application/json' },
      body: JSON.stringify(req),
    }),
  );
}

export async function stopPipeline(): Promise<PipelineStatus> {
  return handle(await fetch(`${API_BASE}/stop`, { method: 'POST' }));
}

export async function getPipelineStatus(): Promise<PipelineStatus> {
  return handle(await fetch(`${API_BASE}/status`));
}

File: src/mrtklib_web_ui/api/clas_pipeline.py
"""API endpoints for the CLAS pipeline (mrtk relay + mrtk cssr2rtcm3)."""

from __future__ import annotations

from fastapi import APIRouter, HTTPException
from pydantic import BaseModel, Field

from mrtklib_web_ui.services import (
    PipelineConfig,
    PipelineState,
    clas_pipeline_service,
)
from mrtklib_web_ui.services.clas_pipeline_service import (
    DEFAULT_BRIDGE_PORT,
    list_serial_ports,
)
from mrtklib_web_ui.services.receiver_presets import list_presets

router = APIRouter()


class PipelineStartRequest(BaseModel):
    receiver_id: str = Field(examples=["septentrio-mosaic-g5"])
    input_device: str = Field(examples=["/dev/ttyUSB0"])
    input_baud: int = Field(default=115200, gt=0)
    bridge_port: int = Field(default=DEFAULT_BRIDGE_PORT, gt=1024, lt=65536)
    sbf_record_path: str | None = Field(
        default=None,
        description=(
            "Optional path under /workspace to record the raw SBF stream "
            "(useful for post-mortem analysis). mrtk relay's time tokens "
            "(%Y, %m, %d, %h, %M, %S) are expanded at write time."
        ),
        examples=["/workspace/clas/sbf_%Y%m%d_%h%M%S.sbf"],
    )


class PipelineStatusResponse(BaseModel):
    state: str
    relay_state: str
    cssr_state: str
    started_at: str | None = None
    error_message: str | None = None
    bridge_port: int | None = None


def _to_status_response(status) -> PipelineStatusResponse:
    return PipelineStatusResponse(
        state=status.state.value,
        relay_state=status.relay_state,
        cssr_state=status.cssr_state,
        started_at=status.started_at.isoformat() if status.started_at else None,
        error_message=status.error_message,
        bridge_port=status.bridge_port,
    )


@router.get("/receivers")
async def get_receivers() -> list[dict]:
    """List supported receiver presets."""
    return list_presets()


@router.get("/serial-ports")
async def get_serial_ports() -> list[dict]:
    """Enumerate serial devices visible to the container."""
    return [{"path": p.path, "label": p.label} for p in list_serial_ports()]


@router.post("/start", response_model=PipelineStatusResponse)
async def start_pipeline(req: PipelineStartRequest) -> PipelineStatusResponse:
    """Start the CLAS pipeline (relay + cssr2rtcm3 + SBF sniffer)."""
    config = PipelineConfig(
        receiver_id=req.receiver_id,
        input_device=req.input_device,
        input_baud=req.input_baud,
        bridge_port=req.bridge_port,
        sbf_record_path=req.sbf_record_path,
    )
    try:
        status = await clas_pipeline_service.start(config)
    except ValueError as e:
        raise HTTPException(status_code=400, detail=str(e))
    except RuntimeError as e:
        raise HTTPException(status_code=409, detail=str(e))

    if status.state == PipelineState.ERROR:
        # Surface device-not-found / device-busy errors with 422 so the UI
        # can show them without treating them as a server bug.
        raise HTTPException(status_code=422, detail=status.error_message or "Failed to start pipeline")
    return _to_status_response(status)


@router.post("/stop", response_model=PipelineStatusResponse)
async def stop_pipeline() -> PipelineStatusResponse:
    """Stop the CLAS pipeline."""
    status = await clas_pipeline_service.stop()
    return _to_status_response(status)


@router.get("/status", response_model=PipelineStatusResponse)
async def get_pipeline_status() -> PipelineStatusResponse:
    """Get current pipeline status."""
    return _to_status_response(clas_pipeline_service.status)
